game-app/game-core/src/main/java/games/strategy/engine/framework/startup/mc/ServerModel.java

```diff
@@ -313,36 +315,58 @@ public synchronized void setLocalPlayerType(final String player, final PlayerTyp
     localPlayerTypes.put(player, type);
   }
 
+  /**
+   * Persists the players mappings to re-use upon a game data change. Used to persist the previous
+   * game's player setting for the game restart if a connection is lost.
+   */
+  public void persistPlayersToNodesMapping() {
+    this.playersToNodesMappingPersisted = true;
+  }
+
+  private void updatePlayersOnGameDataChanged(final GameData data) {
+    // If specified, keep the previous player data.
+    if (playersToNodesMappingPersisted) {
+      playersToNodesMappingPersisted = false;
+      final Set<String> dataPlayers =
+          data.getPlayerList().stream().map(GamePlayer::getName).collect(Collectors.toSet());
+      if (dataPlayers.equals(playersToNodeListing.keySet())) {
+        // List of countries matches, keep the persisted mappings.
```

Comment:
Could you explain a bit more why the list of countries matching is enough to use the 'cached' values for 'playersToNodeListing' etc? Why is it that we know we can skip the additional initialization steps?

From what my (faulty memory) recalls is that we have a lot of parallel data structures that create some very brittle scenarios where if data is one, it is assumed to be in the other. I'm concerned that there could be some room for error there (as we are working with a very brittle design here)

Reply:
The if statement on line 332 is mostly a sanity check.

The expectation is if `playersToNodesMappingPersisted` was set, we're exactly in this "restart" scenario and therefore we should use the persisted data.

The if just guards against some hypothetical code path where we did set that bool, but then the map changed before this code was run. It might actually be impossible - so one option is to make it throw an exception instead.

Reply:
Thanks for the context, I'll want to look at this in more detail. Please allow a few days for more to look at this deeper.

Reply:
Any more thoughts on this? I'm open to us thinking of other approaches long term, but this is still a major improvement over the status quo, so perhaps we can just merge this?

Reply:
Sorry, I have been wanting to take a really deep look at this but have been unable to find the time. 

I wonder, is this code "trivially" correct? (in other words, this is always the case and so these checks are going to be no-ops in all cases) If so, should we remove it? On another side of the coin, is this a good safe guard for data integrity where we should throw an error? On another side, if this is not always the case, is there a nicer way of detecting that we should early exit.

I'm interested in having this PR be merged sooner rather than later. We move forward now, I'm not sure we'll come back. If we could spend some time on these thoughts and try to finish this PR entirely in the next day or two, that would be ideal. I think this is a case where you having write privileges would be ideal @asvitkine  as you could then merge after addressing these comments to the best of your ability.

Reply:
I don't think it's "trivially" correct, because the code is pretty convoluted - so it's hard to tell if this can be hit or not.

On the other hand, my intuition is that it's likely not possible. But I don't have a way to prove that. One option is we go with the exception approach and see if it sticks (e.g. whether it ever gets hit). If it does, then we update the logic.

If it doesn't, we leave it in as a data integrity check to guard against something changing in the future and breaking the functionality. How does that sound?

Reply:
To recap, the plan would be:
- throw an exception
- see if during further prerelease testing we see the problem and then go from there

I can get on board with that. I'm a bit wary that we do enough testing to see the problem or not. If we can get that testing done, then this seems solid.

Assuming we do not 'hit this code', would we remove this check before cutting a release, or maintain it as an exception? It might be nice to do a silent fallback and send an error report, though the complexity of that does not seem worth it to me.

Reply:
I think keeping the exception is helpful because it can help discover a regression if some future refactoring changes the flow such that the assumption no longer holds.